File: mobile/lib/src/data/models/announcement.dart
class Announcement {
  final String id;
  final String title;
  final String content;
  final String type;
  final String targetAudience;
  final List<Attachment> attachments;
  final DateTime publishedAt;
  final DateTime? expiresAt;
  final String status;

  Announcement({
    required this.id,
    required this.title,
    required this.content,
    required this.type,
    required this.targetAudience,
    required this.attachments,
    required this.publishedAt,
    this.expiresAt,
    required this.status,
  });

  bool get isExpired => expiresAt != null && DateTime.now().isAfter(expiresAt!);

  factory Announcement.fromJson(Map<String, dynamic> json) {
    return Announcement(
      id: json['_id'] ?? json['id'],
      title: json['title'],
      content: json['content'],
      type: json['type'],
      targetAudience: json['targetAudience'],
      attachments:
          (json['attachments'] as List?)
              ?.map((a) => Attachment.fromJson(a))
              .toList() ??
          [],
      publishedAt: DateTime.parse(json['publishedAt']),
      expiresAt: json['expiresAt'] != null
          ? DateTime.parse(json['expiresAt'])
          : null,
      status: json['status'],
    );
  }
}

class Attachment {
  final String filename;
  final String url;
  final int? size;

  Attachment({required this.filename, required this.url, this.size});

  factory Attachment.fromJson(Map<String, dynamic> json) {
    return Attachment(
      filename: json['filename'],
      url: json['url'],
      size: json['size'],
    );
  }
}

File: server/src/models/Announcement.js
const mongoose = require('mongoose');

const announcementSchema = new mongoose.Schema({
    title: {
        type: String,
        required: true
    },
    content: {
        type: String,
        required: true
    },
    type: {
        type: String,
        enum: ['general', 'urgent', 'event', 'holiday', 'exam'],
        default: 'general'
    },
    targetAudience: {
        type: String,
        enum: ['all', 'students', 'teachers', 'parents', 'specific_class'],
        default: 'all'
    },
    targetClasses: [{
        class: String,
        section: String
    }],
    createdBy: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
        required: true
    },
    attachments: [{
        filename: String,
        url: String,
        size: Number
    }],
    publishedAt: {
        type: Date,
        default: Date.now
    },
    expiresAt: Date,
    sendNotification: {
        type: Boolean,
        default: true
    },
    notificationSent: {
        type: Boolean,
        default: false
    },
    status: {
        type: String,
        enum: ['draft', 'published', 'archived'],
        default: 'published'
    }
}, {
    timestamps: true
});

// Index for querying announcements
announcementSchema.index({ publishedAt: -1 });
announcementSchema.index({ status: 1, publishedAt: -1 });

module.exports = mongoose.model('Announcement', announcementSchema);
